hazelcast/src/main/java/com/hazelcast/sql/impl/exec/root/BlockingRootResultConsumer.java

```diff
@@ -34,27 +33,32 @@
     /** Iterator over produced rows. */
     private final InternalIterator iterator = new InternalIterator();
 
-    /** Query context to schedule root execution when the next batch is needed. */
-    private volatile QueryFragmentContext context;
+    /** A callback to schedule root execution when the next batch is needed. */
+    private volatile ScheduleCallback scheduleCallback;
 
     /** The batch that is currently being consumed. */
     private List<Row> currentBatch;
 
-    /** When "true" no more batches are expected. */
+    /** When "true", no more batches are expected. */
     private boolean done;
 
     /** Error which occurred during query execution. */
     private QueryException doneError;
 
     @Override
-    public void setup(QueryFragmentContext context) {
-        this.context = context;
+    public void setup(ScheduleCallback scheduleCallback) {
+        this.scheduleCallback = scheduleCallback;
     }
 
     @Override
     public boolean consume(List<Row> batch, boolean last) {
         synchronized (mux) {
-            assert !done;
+            if (done) {
+                // this is possible if the query was concurrently cancelled
+                // see https://github.com/hazelcast/hazelcast/issues/17160
+                assert doneError != null;
+                throw new RuntimeException(doneError);
```

Comment:
Why do we throw a new `RuntimeException` here? Will `throw doneError;` be a proper to address #17160?

Reply:
We wrap in another `RuntimeException` to have a proper stack trace: we want a stack trace from the place where the exception was thrown, not where it was created. This exception is not propagated to the user, it is used to tear down the jet job after a cancellation and will be logged on the member. It also cancels the scheduled imdg task.

Yes, it's a fix to #17160.

Reply:
Is this condition specific to Jet only?

Reply:
Nope, it's a general principle. Without this wrapping the stack trace will point you somewhere completely else than where the exception was actually thrown.

Reply:
I do not quite get how it plays with the IMDG engine. First question - why we choose `RuntimeException`, while in all other parts of the engine we work with `QueryException` that carries additional information, which is eventually converted to public `SqlException`? 

The second question is more important - why should we rethrow an exception at all? The idea of `doneError` is that this is an exception that is propagated to the user eventually. But `consume` is an internal method. From the IMDG perspective, there is no sense to throw an exception there, because the error is already reported, so the executor will be terminate soon anyway, and throwing another exception from `consume` will do nothing. Why do we need to throw an exception from here for Jet?

Reply:
Reverted that particular change

Reply:
But now we hit the problem with the wrong assert :-) What I meant is that for the IMDG engine, throwing error from here is basically no-op because the other error has already been installed. So the following code is most likely to work just fine:
```
public boolean consume(List<Row> batch, boolean last) {
    synchronized (mux) {
        if (done) {
            return false;
        }
        ...
    }  
}
```
In this case, the `RootExec` will return `WAIT` and deactivate the fragment. This is ok because the original error already triggered cancel operation, which will clear the fragment eventually. Re-throwing the error here makes no difference because the original error is already installed, and the new error will be ignored in `QueryState.cancel` when doing `CAS(null, error)`.

The question is whether Jet requires throwing an error from `consume`?

Reply:
I feel like addressing this issue should be part of a separate PR

Reply:
I consider copying this whole class. It's tailored for IMDG needs, it's not a well-defined reusable API. For example, Jet doesn't need the `setup` or the callback. It's prone to incompatible changes in the future.

Then the issue #17160 remains to be fixed by imdg guys.